### PR TITLE
Add several CAST translations and tests

### PR DESCRIPTION
**Pull request by LABSARI**
Add translations for :
- TMethod
- cppIfTrue: as argument
- cCode: as argument

Add tests for :
- Method comments
- cppIfTrue: as argument
- cCode: as argument

It also fixes some printing in Slang and CAST.

### CODE REVIEW

File: smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st [hunk 1]
```diff
@@ -1618,6 +1675,187 @@ SlangBasicTranslationTest >> testSendCppIfIfTrueIfFalse [
...
+							          expression: (TConstantNode value: false)).
+						         (TAssignmentNode new
+							          setVariable: (TVariableNode new setName: 'c')
+							          expression: (TConstantNode value: 1)) }) }.
+	translation := self translate: send.
+
+	self assert: translation trimBoth equals: '{
+	a = 0;
+	c = 1;
+}'
```
Comment (LABSARI) [inline smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st:2154]: This is weird. If you look at the comment, the part "true ifTrue: [ false ] ifFalse: [ true ]" is evaluated as false, but Slang checks (through CCodeGenerator>>nilOrBooleanConditionFor:) if the receiver (which is a true constant) of the ifTrue:ifFalse: message is a boolean constant. Why the receiver ? It shouldn't print anything in this case, right ?

Comment (guillep) [reply smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st:2154]: Let's give a quick look tomorrow. I'd say you've found a bug :)

File: smalltalksrc/Slang-Tests/CSlangPrettyPrinterTest.class.st [hunk 2]
```diff
@@ -320,6 +320,26 @@ CSlangPrettyPrinterTest >> testVisitForStatement [
 }'
 ]
 
+{ #category : #'tests-visitor' }
+CSlangPrettyPrinterTest >> testVisitFunctionDefinition [
```
Comment (guillep) [inline smalltalksrc/Slang-Tests/CSlangPrettyPrinterTest.class.st:324]: Maybe we should add a comment to explain that the cDeclaration is a raw declaration and it wins over a normal declarator.
And name the test to make it more explicit

File: smalltalksrc/Slang-Tests/CSlangPrettyPrinterTest.class.st [hunk 3]
```diff
@@ -414,6 +434,17 @@ CSlangPrettyPrinterTest >> testVisitInitializer [
 
 ]
 
+{ #category : #'tests-visitor' }
+CSlangPrettyPrinterTest >> testVisitParameterDeclaration [
+
+	| print |
+	print := self prettyPrint: ((CParameterDeclarationNode
```
Comment (guillep) [inline smalltalksrc/Slang-Tests/CSlangPrettyPrinterTest.class.st:441]: Same as above

File: smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st [hunk 4]
```diff
@@ -875,6 +874,34 @@ SlangBasicTranslationTest >> testGoto [
 	self assert: translation equals: 'goto lab'
 ]
 
+{ #category : #'tests-method' }
+SlangBasicTranslationTest >> testMethodComment [
+
+	""
```
Comment (guillep) [inline smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st]: Empty comment?

File: smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st [hunk 5]
```diff
@@ -875,6 +874,34 @@ SlangBasicTranslationTest >> testGoto [
...
+	| translation tMethod |
+	generator addClass: SlangBasicTranslationTestClass.
+	tMethod := generator methodNamed: #foo.
```
Comment (guillep) [inline smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st]: maybe name foo after what it does. If empty, it could be called emptyMethod

Comment (guillep) [reply smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st]: Actually, is a method with a comment. So maybe call it methodWithMultilineComment

File: smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st [hunk 6]
```diff
@@ -875,6 +874,34 @@ SlangBasicTranslationTest >> testGoto [
...
+]
+
+{ #category : #'tests-method' }
+SlangBasicTranslationTest >> testMethodCommentOriginalMethod [
+
+	"Comment that includes the pharo style method signature : ClassName>>MethodName "
+	| translation tMethod |
+	generator addClass: SlangBasicTranslationTestClass.
+	tMethod := generator methodNamed: #foo.
+	translation := self translate: tMethod.
```
Comment (guillep) [inline smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st:1126]: Maybe extract these three lines into a helper method

```smalltalk
translation := self translateMethod: #foo.
```

File: smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st [hunk 7]
```diff
@@ -1618,6 +1675,187 @@ SlangBasicTranslationTest >> testSendCppIfIfTrueIfFalse [
 #endif /* FEATURE */'
 ]
 
+{ #category : #'tests-builtins' }
+SlangBasicTranslationTest >> testSendCppIfIfTrueIfFalseAsArgument [
+
+	"A C preprocessor condition can anywhere in code, even in an expression statement."
```
Comment (guillep) [inline smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st:1979]: Put the original code as a comment maybe

File: smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st [hunk 8]
```diff
@@ -1618,6 +1675,187 @@ SlangBasicTranslationTest >> testSendCppIfIfTrueIfFalse [
...
+						          setSelector: #cppIf:ifTrue:ifFalse:
+						          receiver: (TVariableNode new setName: 'self')
+						          arguments: { 
+								          (TConstantNode value: #FEATURE).
+								          (TConstantNode value: 2).
+								          (TConstantNode value: 4) }) }).
+	translation := self translate: send.
+
+	self assert: translation trimBoth equals: 'y = ((x < 10)
+	 ?
```
Comment (guillep) [inline smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st:2000]: Can this test case be simplified removing the iftrue from the middle?

File: smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st [hunk 9]
```diff
@@ -1618,6 +1675,187 @@ SlangBasicTranslationTest >> testSendCppIfIfTrueIfFalse [
...
+							           receiver: (TVariableNode new setName: 'self')
+							           arguments: { 
+									           (TConstantNode value: #FEATURE).
+									           (TConstantNode value: 2).
+									           (TConstantNode value: 4) }) })) }.
+	translation := self translate: send.
+
+	self assert: translation trimBoth equals: '{
+	y = ((x < 10)
+		 ?
```
Comment (guillep) [inline smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st:2035]: Same as above

File: smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st [hunk 10]
```diff
@@ -1618,6 +1675,187 @@ SlangBasicTranslationTest >> testSendCppIfIfTrueIfFalse [
...
+	{
+		a = 1;
+		c = 0;
+	}
+#  endif /* FEATURE */
+}'
+]
+
+{ #category : #'tests-builtins' }
+SlangBasicTranslationTest >> testSendCppIfIfTrueIndentationInBlock [
```
Comment (guillep) [inline smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st]: Nested blocks indentation

File: smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st [hunk 11]
```diff
@@ -1618,6 +1675,187 @@ SlangBasicTranslationTest >> testSendCppIfIfTrueIfFalse [
...
+		}
+#    endif /* FEATURE */
+	}
+}'
+]
+
+{ #category : #'tests-builtins' }
+SlangBasicTranslationTest >> testSendCppIfIfTrueWithSendIfTrueAsCondition [
+
+	"self cppIf: (true ifTrue: [ false ] ifFalse: [ true ]) ifTrue: [ a := false. c := 1 ]"
```
Comment (guillep) [inline smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st:2126]: `self cppIf: false ifTrue: [ a := false. c := 1 ]` ?

File: smalltalksrc/Slang/CSlangPrettyPrinter.class.st [hunk 12]
```diff
@@ -169,14 +170,14 @@ CSlangPrettyPrinter >> visitConstant: aConstant [
...
-			stream space ].
-	
-	aDeclaration declarators
-		do: [ :e | e acceptVisitor: self ]
-		separatedBy: [ stream nextPutAll: ', ' ].
+	aDeclaration specifiers do: [ :e | 
+		stream nextPutAll: e.
+		stream space ].
+
+	aDeclaration hasRawPrototype ifFalse: [
```
Comment (guillep) [inline smalltalksrc/Slang/CSlangPrettyPrinter.class.st:188]: Maybe use a different class for this

File: smalltalksrc/Slang/CSlangPrettyPrinter.class.st [hunk 13]
```diff
@@ -246,6 +247,30 @@ CSlangPrettyPrinter >> visitForStatement: aForStatement [
...
+]
+
+{ #category : #generated }
+CSlangPrettyPrinter >> visitFunctionDefinition: aFunction [
+
+	aFunction specifiers
+		do: [ :e | stream nextPutAll: e ]
+		separatedBy: [ stream space ].
+	stream cr.
+	aFunction hasRawPrototype ifFalse: [
```
Comment (guillep) [inline smalltalksrc/Slang/CSlangPrettyPrinter.class.st:279]: Same as above

File: smalltalksrc/Slang/CSlangPrettyPrinter.class.st [hunk 14]
```diff
@@ -317,6 +347,17 @@ CSlangPrettyPrinter >> visitLabeledStatement: aLabeledStatement [
...
 
+{ #category : #generated }
+CSlangPrettyPrinter >> visitParameterDeclaration: aParameter [
+
+	"aParameter specifiers
+		do: [ :e | stream nextPutAll: e ]
+		separatedBy: [ stream space ].
+	stream space.
+	aParameter declarator acceptVisitor: self"
+	stream nextPutAll: aParameter rawDeclaration
```
Comment (guillep) [inline smalltalksrc/Slang/CSlangPrettyPrinter.class.st]: Maybe there are missing tests about parameters without raw declarations

**Review by guillep — APPROVED** [review]

**Review by guillep — APPROVED** [review]
Good! Let's wait for the tests.

I've put all comments as new tasks in the trello.

**Comment by guillep** [comment]
There is a merge conflict. I'll resolve it in my machine and push the merge from there

**Comment by guillep** [comment]
Iceberg knew how to do the merge automatically :). Let's wait for the tests now

**Comment by guillep** [comment]
And locally everything is green ;) 
image

**Comment by guillep** [comment]
[cppIf:ifTrue:](https://trello.com/c/KxbPOwsJ/17-cppififtrue)

**Comment by guillep** [comment]
[Translation of conditional compilation of methods](https://trello.com/c/Mi46ZGjo/58-translation-of-conditional-compilation-of-methods)

**Comment by guillep** [comment]
[Translation of variable declarations in methods](https://trello.com/c/epALT0iM/57-translation-of-variable-declarations-in-methods)

**Comment by guillep** [comment]
[Translation of Function Prototypes for Method Translation](https://trello.com/c/v5SwISRI/56-translation-of-function-prototypes-for-method-translation)

**Comment by guillep** [comment]
[Translation of method comments](https://trello.com/c/KebzXayD/59-translation-of-method-comments)

**Comment by guillep** [comment]
[cCode:](https://trello.com/c/I5dNrpCH/14-ccode)

**Pull request merged**